src/servers/io_test_util.cc

```
@@ -0,0 +1,697 @@
+                << cudaGetErrorString(err);
+    }
+  }
+};
+
+void
+Usage(char** argv, const std::string& msg = std::string())
+{
+  if (!msg.empty()) {
+    LOG_ERROR << msg;
+  }
+
+  LOG_ERROR << "Usage: " << argv[0] << " [options]";
+  LOG_ERROR << "\t-i [input device ID]";
+  LOG_ERROR << "\t-out [output device ID]";
+  LOG_ERROR << "\t-v Enable verbose logging";
+  LOG_ERROR << "\t-r [model repository absolute path]";
+  LOG_ERROR << "\t-m [model name to be tested]";
+  LOG_ERROR << "\tFor device ID, -1 is used to stand for CPU device, "
+            << "non-negative value is for GPU device.";
+
+  exit(1);
+}
+
+std::string
+MemoryTypeString(TRTSERVER_Memory_Type memory_type)
+{
+  return (memory_type == TRTSERVER_MEMORY_CPU) ? "CPU memory" : "GPU memory";
+}
+
+TRTSERVER_Error*
+ResponseAlloc(
+    TRTSERVER_ResponseAllocator* allocator, void** buffer, void** buffer_userp,
+    const char* tensor_name, size_t byte_size,
+    TRTSERVER_Memory_Type memory_type, int64_t memory_type_id, void* userp)
+{
+  // Pass the tensor name with buffer_userp so we can show it when
+  // releasing the buffer.
+
+  // If 'byte_size' is zero just return 'buffer'==nullptr, we don't
+  // need to do any other book-keeping.
+  if (byte_size == 0) {
+    *buffer = nullptr;
+    *buffer_userp = nullptr;
+    LOG_INFO << "allocated " << byte_size << " bytes for result tensor "
+             << tensor_name;
+  } else {
+    void* allocated_ptr = nullptr;
+    if (memory_type == TRTSERVER_MEMORY_CPU) {
+      allocated_ptr = malloc(byte_size);
+    } else if (io_spec.output_type_ == TRTSERVER_MEMORY_GPU) {
+      auto err = cudaSetDevice(memory_type_id);
+      if (err == cudaSuccess) {
+        err = cudaMalloc(&allocated_ptr, byte_size);
+      }
+      if (err != cudaSuccess) {
+        LOG_INFO << "cudaMalloc failed: " << cudaGetErrorString(err);
+        allocated_ptr = nullptr;
+      }
+    }
+
+    if (allocated_ptr != nullptr) {
```

If allocated_ptr is nullptr here it means allocation failed so should return an appropriate error code.